### PR TITLE
refactor: use more gsl::not_null in utils.h and deterministicmns.h

### PR DESCRIPTION
**Pull request by PastaPastaPasta**
## Issue being fixed or feature implemented
Use not_null if the function would crash if given a nullptr

## What was done?
Refactored to use gsl::not_null

## How Has This Been Tested?
Compiled

## Breaking Changes
Should be none

## Checklist:
  _Go over all the following points, and put an `x` in all the boxes that apply._
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone _(for repository code-owners and collaborators only)_

### CODE REVIEW

File: src/evo/deterministicmns.h [hunk 1]
```diff
@@ -420,9 +421,7 @@ class CDeterministicMNList
     template <typename T>
     [[nodiscard]] uint256 GetUniquePropertyHash(const T& v) const
     {
-        if constexpr (std::is_same<T, CBLSPublicKey>()) {
-            assert(false);
-        }
+        static_assert(!std::is_same<T, CBLSPublicKey>(), "GetUniquePropertyHash cannot be templated against CBLSPublicKey");
```
Comment (UdjinM6) [inline src/evo/deterministicmns.h:424]: nit: unrelated

File: src/evo/deterministicmns.cpp [hunk 2]
```diff
@@ -1667,50 +1667,48 @@ bool CheckProUpServTx(const CTransaction& tx, const CBlockIndex* pindexPrev, TxV
         }
     }
 
-    if (pindexPrev) {
```
Comment (UdjinM6) [inline src/evo/deterministicmns.cpp:1670]: Strictly saying, this is a change in behaviour and not a refactor (same for other `CheckPro...Tx`s). But it doesn't crash here as far as I see so ok.

File: src/llmq/utils.cpp [hunk 3]
```diff
@@ -1019,7 +1009,7 @@ bool IsQuorumTypeEnabledInternal(Consensus::LLMQType llmqType, const CQuorumMana
             return VersionBitsState(pindex, consensusParams, Consensus::DEPLOYMENT_TESTDUMMY, llmq_versionbitscache) == ThresholdState::ACTIVE;
         }
         case Consensus::LLMQType::LLMQ_100_67:
-            return pindex != nullptr && pindex->nHeight + 1 >= consensusParams.DIP0020Height;
+            return pindex->nHeight + 1 >= consensusParams.DIP0020Height;
```
Comment (UdjinM6) [inline src/llmq/utils.cpp:1012]: same, a change in behaviour here but seems to be working.

**Review by UdjinM6 — APPROVED** [review]
utACK (with a few notes)

**Comment by knst** [comment]
> Conflicting PRs:  [5590]

Please, review #5590  first

File: src/llmq/utils.cpp (hunk 3 above)
Comment (PastaPastaPasta) [reply src/llmq/utils.cpp:1012]: Other switch cases fail if nullptr so seems acceptable.

File: src/evo/deterministicmns.cpp (hunk 2 above)
Comment (PastaPastaPasta) [reply src/evo/deterministicmns.cpp:1670]: Not really as the function would crash earlier.

File: src/llmq/utils.h [hunk 4]
```diff
@@ -46,38 +47,38 @@ namespace utils
 {
 
 // includes members which failed DKG
-std::vector<CDeterministicMNCPtr> GetAllQuorumMembers(Consensus::LLMQType llmqType, const CBlockIndex* pQuorumBaseBlockIndex, bool reset_cache = false);
+std::vector<CDeterministicMNCPtr> GetAllQuorumMembers(Consensus::LLMQType llmqType, gsl::not_null<const CBlockIndex*> pQuorumBaseBlockIndex, bool reset_cache = false);
```
Comment (knst) [inline src/llmq/utils.h:50]: so far as this and many other places requires not-null anyway, why even don't change `const CBlockIndex*> to the `const CBlockIndex&` ?

For reference, couple PR with bitcoin's similar changes of `CBlockIndex*` to `CBlockIndex&`. Bitcoin doesn't change it everywhere, but only when it is obviously non-nullptr, such as:
https://github.com/bitcoin/bitcoin/pull/25677/files
https://github.com/bitcoin/bitcoin/pull/25016/files

Comment (PastaPastaPasta) [reply src/llmq/utils.h:50]: Well; so one benefit imo is that dereferencing a nullptr is undefined behavior not guaranteed to be caught at runtime; additionally, gsl::not_null doesn't require additional syntax to pass to a later func that needs a ptr or additional syntax when passing in a ptr. In bitcoin PR you mention it appears they force to use something like `*Assert(m_chain.Tip()` which is significantly more typing :D

Comment (knst) [reply src/llmq/utils.h:50]: well, `*Assert(...)` is 7 bytes shorter than `gsl::not_null<...>(...)`

And indeed, you can avoid UB by validating it before converting to ference.

Comment (knst) [reply src/llmq/utils.h:50]: yeah, but once you convert pointer to reference, you can assume that all recursive calls; submethods; helpers, etc are never have nullptr anymore; but for case of pointer - you need to add gsl:not_null to every method do show that is not nullptr.

Comment (PastaPastaPasta) [reply src/llmq/utils.h:50]: Well yes; same as every method needs to be declared as a reference. Once a not_null is constructed it is zero cost to pass it to another not_null function

**Review by knst — APPROVED** [review]
utACK, 

`gsl::not_null` actually looks good for most of usages in this PR

**Review by UdjinM6 — APPROVED** [review]
utACK

**Review by knst — APPROVED** [review]
utACK
